File: me_graveler.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

const int ATTEMPTS = 1000000000;
int rolls = ATTEMPTS;
int maxOnes = 0;
int score;
int number;

int main(void){
    clock_t start_time, end_time;
    double cpu_time_used;

    srand(time(NULL));

    start_time = clock();
    for(int i = 0; i < ATTEMPTS; i++){
        if(maxOnes >= 177){
            rolls = i;
            break;
        }
        score = 0;
        for (int j = 0; j < 231; ){
            number =  rand()*rand(); //rand is taking 2.5 seconds per million iterations
            
            while (number>4){
                score+= (number%4)==1;
                number = number / 4;
                j++;
                if (j == 231){
                    break;
                }
            }
            if ((maxOnes-score)>(231-j)){
                break;
            }
        }
        if(score > maxOnes){
            maxOnes = score;
        }
    }

    end_time = clock();
    cpu_time_used = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;

    printf("highest roll: %d \n", maxOnes);
    printf("number of rolls: %d \n", rolls);
    printf("time taken: %f \n", cpu_time_used);

    return 0;
}

/*
TIME TAKEN FOR 1,000,000,000 656.576 seconds.
TIME TAKEN FOR 1,000,000 0.702 seconds
*/

File: me_graveler.py
from random import random
import time



#total number of attempts the (try to get one billion down to a reasonable time)
attempts = 1_000_000
rolls = attempts
maxOnes = 0
start_time = time.time()
# Using a for loop because it is faster than a while loop
for i in range(attempts):
    # takes 0.3 seconds per million attempts
    if maxOnes >= 177:
        rolls = i
        break
    score = 0
    for j in range(231):
       # takes 7 seconds per 1,000,000 attempts
        if random()<=0.25:
            score += 1
        
    #takes about 0.1 seconds per million attempts
    maxOnes += (score * (score > maxOnes)) - (maxOnes * (score > maxOnes))
    

end_time = time.time()
elapsed_time = end_time - start_time

print("Highest Ones Roll:",maxOnes)
print("Number of Roll Sessions:",rolls)
print("time taken:", elapsed_time)

"""
TIME TAKEN FOR 1,000,000 10.267 seconds
"""
